server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -635,9 +635,8 @@ private void doDeleteShardSnapshots(Collection<SnapshotId> snapshotIds, long rep
                 for (ShardSnapshotMetaDeleteResult newGen : deleteResults) {
                     builder.put(newGen.indexId, newGen.shardId, newGen.newGeneration);
                 }
-                final RepositoryData updatedRepoData = repositoryData.removeSnapshots(snapshotIds, builder.build());
-                writeIndexGen(updatedRepoData, repositoryStateId, repoMetaVersion, Function.identity(),
-                    ActionListener.wrap(v -> writeUpdatedRepoDataStep.onResponse(updatedRepoData), listener::onFailure));
+                writeIndexGen(repositoryData.removeSnapshots(snapshotIds, builder.build()), repositoryStateId, repoMetaVersion,
+                        Function.identity(), writeUpdatedRepoDataStep);
```

Comment:
We were passing repo data without the correct generation set on it to the listener here. Also the indirection to `listener::onFailure`